frontend/catalyst/jax_primitives.py

```diff
@@ -368,6 +403,133 @@ def _print_lowering(jax_ctx: mlir.LoweringRuleContext, *args, string=None, memre
+def _transform_named_sequence_p_abstract_eval(*args):
+    return ()
+
+
+@transform_named_sequence_p.def_impl
+def _transform_named_sequence_p_def_impl(*args):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _transform_named_sequence_lowering(jax_ctx: mlir.LoweringRuleContext, *args):
+    module = jax_ctx.module_context.module
+
+    # If there already is a transform.named_sequence in the module, don't add another one!
+    # Do nothing and exit!
+    for i in range(len(module.body.operations)):
+        if module.body.operations[i].operation.name == "transform.named_sequence":
+            return module.body.operations[i].operation.results
+
+    functype = ir.FunctionType.get(inputs=[transform_func_type], results=[])
+    functype_attr = ir.TypeAttr.get(functype)
+
+    # add the transform.with_named_sequence unit attr to the parent module
+    with_named_sequence_attr = ir.UnitAttr.get(jax_ctx.module_context.context)
+    module.operation.attributes["transform.with_named_sequence"] = with_named_sequence_attr
+
+    # Insert the transform.named_sequence op into the module to be transformed
+    # Note that InsertionPoint(Block) inserts after the last operation but still inside the block.
+    # Hence we insert at the end, inside the module
+    with ir.InsertionPoint(module.body):
+        named_sequence_op = NamedSequenceOp(
+            sym_name="__transform_main",
+            function_type=functype_attr,
+        )
+
+        # transform.named_sequence op is the "main function" of the transform dialect
+        # and thus needs an entry block (which also should be its only block)
+        # The argument of the block is the target function to run the transform sequence on
+        bb = ir.Block.create_at_start(named_sequence_op.body, arg_types=[transform_func_type])
+
+        # The transform.named_sequence needs a terminator called "transform.yield"
+        with ir.InsertionPoint(bb):
+            transform_yield_op = TransformYieldOp(operands_=[])  # pylint: disable=unused-variable
+
+    return named_sequence_op.results
+
+
+#
+# apply_registered_pass
+#
+@apply_registered_pass_p.def_abstract_eval
+def _apply_registered_pass_abstract_eval(*args, pass_name, options=None):
+    return AbstractTransformFunc()
+
+
+@apply_registered_pass_p.def_impl
+def _apply_registered_pass_def_impl(*args, pass_name, options=None):  # pragma: no cover
+    raise NotImplementedError()
+
+
+def _apply_registered_pass_lowering(
+    jax_ctx: mlir.LoweringRuleContext, *args, pass_name, options=None
+):
+    module = jax_ctx.module_context.module
+    named_sequence_op = None
+    for i in reversed(range(len(module.body.operations))):
+        # transform.named_sequence usually is at the end of the module, so look for it from the end
+        if module.body.operations[i].operation.name == "transform.named_sequence":
+            named_sequence_op = module.body.operations[i].operation
+            break
```

Comment:
You might be able to walk the MLIR using some sort of interface. We haven't had a need for it, so we've never used it, but I believe it may be available.

Reply:
https://github.com/llvm/llvm-project/pull/87962 I guess it is relatively new :sweat_smile: 

```
def testOpWalk():
    ctx = Context()
    ctx.allow_unregistered_dialects = True
    module = Module.parse(
        r"""
    builtin.module {
      func.func @f() {
        func.return
      }
    }
  """,
        ctx,
    )

    def callback(op):
        print(op.name)
        return WalkResult.ADVANCE

    # Test post-order walk (default).
    # CHECK-NEXT:  Post-order
    # CHECK-NEXT:  func.return
    # CHECK-NEXT:  func.func
    # CHECK-NEXT:  builtin.module
    print("Post-order")
    module.operation.walk(callback)
```

Reply:
Oh dope :D

Reply:
any action item here if we don't see this llvm commit?

Reply:
None from my side :)